### ISSUE TITLE
Super interfaces are not parsed in the ITree model

### ISSUE THREAD

**Issue by algomaster99** [SpoonLabs/gumtree-spoon-ast-diff#181]
The super interfaces are ignored for now because instances of `CtReference` in general are ignored. Consider the code,
```java
class Example implements A, B, C { }
```
Its tree only has two nodes - `root` and `Example`.
```
-1@@
	65190232@@Example
```

I propose two models via which super interfaces can be incorporated.

- ```
  -1@@
	65190232@@Example
              <int>@@A
              <int>@@B
              <int>@@C
              ...other children...
  ```
- ```
  -1@@
	65190232@@Example
              <int>@@SUPER_INTERFACE
                   <int>@@A
                   <int>@@B
                   <int>@@C
              ...other children...
  ```

I am more inclined towards the second approach because of semantic reasons. Moreover, modifiers are also clubbed into a single parent called a virtual element.

**Comment by monperrus** [comment]
Thanks for the bug analysis.

> I am more inclined towards the second approach because of semantic reasons.

agree

**Comment by slarse** [comment]
> > I am more inclined towards the second approach because of semantic reasons.
>
> agree

agree with the agree

**Comment by algomaster99** [comment]
I started working on this issue and realised that the second approach will require extra methods to parse nested types of interfaces; something similar to [this](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/pull/179/files#diff-19ac4c8d46440718985d3a70cfd1e2eff828f1cdbeb99bc306993a5783e6b3abR79-R88). I don't support this recursive way because we would be restricted to adding only type arguments of `CtTypeReference` in the tree (not sure there if there can be some other type of child of `CtTypeReference`). Moreover, `TreeScanner` is literally doing the same thing and I feel that we should tweak it instead of adding such methods.

Before explaining the approaches, let me explain how are nodes appended in the tree. This would help us understand when we can add children to the desired node.

There are mainly two methods governing this:
1. [`pushNodeToTree`](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/master/src/main/java/gumtree/spoon/builder/TreeScanner.java#L94): This adds a child node to the topmost element in the stack (`nodes`) and then pushes the child node onto the stack. Moreover, this is a private method and can only be used in the class - `TreeScanner`. Whenever an element `enter`s `TreeScanner`, this method is called upon that element.
2. [`addSiblingNode`](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/master/src/main/java/gumtree/spoon/builder/TreeScanner.java#L102): Only adds a child node to the topmost element in the stack.

### Approach 1

Inserted a condition in `isToIgnore` to return `false` when the role of an element in its parent is `INTERFACE`. This returned a tree exactly like above. Since each super interface will be a topmost element in the stack, we will be able to append children (nested types) using `TreeScanner` itself.

### Approach 2

I enter `NodeCreator` after pushing the class (`Example`) node in `TreeScanner`. Then I override [`scanTypeInformation`](https://github.com/INRIA/spoon/blob/master/src/main/java/spoon/reflect/visitor/CtInheritanceScanner.java#L510) - a method visited by the `InheritanceScanner`.

```java
@Override
public <T> void scanCtTypeInformation(CtTypeInformation typeInfo) {
	if (typeInfo.getSuperInterfaces().isEmpty()) {
		return;
	}
	ITree superInterfaceRoot = builder.createNode("INTERFACE", "SuperInterfaces");
	for (CtTypeReference<?> type: typeInfo.getSuperInterfaces()) {
		ITree superInterfaceNode = builder.createNode(getClassName(type.getClass().getSimpleName()), type.getQualifiedName());
		superInterfaceNode.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, type);
		type.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, superInterfaceNode);
		superInterfaceRoot.addChild(superInterfaceNode);
	}
	builder.addSiblingNode(superInterfaceRoot);
}
```

This adds child, `SUPER_INTERFACE`, to the `Example` node and then proceeds to add each super interface as a child to the `SUPER_INTERFACE` node. This results in a tree exactly like shown above in the second approach.

Now, if we want to add children of each super interface, we would have to do something like [this](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/pull/179/files#diff-19ac4c8d46440718985d3a70cfd1e2eff828f1cdbeb99bc306993a5783e6b3abR79-R88) inside the for-loop because we cannot use `TreeScanner` to add children as the super interfaces will never be the topmost element in the stack, hence cannot be accessed by `pushNodeToTree`. Moreover, we cannot append to stack inside `NodeCreator` because it will [break the contract](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/master/src/main/java/gumtree/spoon/builder/TreeScanner.java#L52).

**Comment by algomaster99** [comment]
This comment is just for more context.

While I am dealing with the latest issues, I keep in mind that the patch is coherent with my future fix of #173 . Since there are two scanners, `TreeScanner` and `NodeCreator`, running simultaneously, there are always two ways to add a node in an `ITree` instance and as of now, I am still finding which scanner would be best for #173 using  different issues (like this one).

Using `TreeScanner` may add a lot of `if-else` conditions in `isToIgnore` making it cluttered meanwhile using `NodeCreator` may require me to override many `scanCt*` methods so I am trying to strike a balance.
